src/doc/src/reference/features-examples.md

```diff
@@ -0,0 +1,177 @@
+## Features Examples
+
+The following illustrates some real-world examples of features in action.
+
+### Minimizing build times and file sizes
+
+Some packages use features so that if the features are not enabled, it reduces
+the size of the crate and reduces compile time. Some examples are:
+
+* [`syn`] is a popular crate for parsing Rust code. Since it is so popular, it
+  is helpful to reduce compile times since it affects so many projects. It has
+  a [clearly documented list][syn-features] of features which can be used to
+  minimize the amount of code it contains.
+* [`winapi`] has [a large number][winapi-features] of features that
+  limit which Windows API bindings it supports.
+* [`regex`] has a [several features][regex-features] that are [well
+  documented][regex-docs]. Cutting out Unicode support can reduce the
+  resulting file size as it can remove some large tables.
+
+[`winapi`]: https://crates.io/crates/winapi
+[winapi-features]: https://github.com/retep998/winapi-rs/blob/0.3.9/Cargo.toml#L25-L431
+[`regex`]: https://crates.io/crates/regex
+[`syn`]: https://crates.io/crates/syn
+[syn-features]: https://docs.rs/syn/1.0.54/syn/#optional-features
+[regex-features]: https://github.com/rust-lang/regex/blob/1.4.2/Cargo.toml#L33-L101
+[regex-docs]: https://docs.rs/regex/1.4.2/regex/#crate-features
+
+### Extending behavior
+
+The [`serde_json`] package has a [`preserve_order` feature][serde_json-preserve_order]
+which [changes the behavior][serde_json-code] of JSON maps to preserve the
+order that keys are inserted. Notice that it enables an optional dependency
+[`indexmap`] to implement the new behavior.
+
+When changing behavior like this, be careful to make sure the changes are
+[SemVer compatible]. That is, enabling the feature should not break code that
+usually builds with the feature off.
+
+[`serde_json`]: https://crates.io/crates/serde_json
+[serde_json-preserve_order]: https://github.com/serde-rs/json/blob/v1.0.60/Cargo.toml#L53-L56
+[SemVer compatible]: features.md#semver-compatibility
+[serde_json-code]: https://github.com/serde-rs/json/blob/v1.0.60/src/map.rs#L23-L26
+[`indexmap`]: https://crates.io/crates/indexmap
+
+### `no_std` support
+
+Some packages want to support both [`no_std`] and `std` environments. This is
+useful for supporting embedded and resource-constrained platforms, but still
+allowing extended capabilities for platforms that support the full standard
+library. The [`memchr`] package defines a [`std` feature] that is [enabled by
+default][memchr-default]. At the top of the library, it [conditionally enables
+the `no_std` attribute][memchr-nostd]. Then, in various places in the code, it
```

Comment:
Technically I would not recommend this design to others (a conditional `#![no_std]` attribute), because it's generally easier to develop libraries if they unconditionally define the attribute and conditionally import `std` based on features. Then if items from `std` are desired they are imported in the relevant locations with `use std::...` (etc)